tests/testthat/test-return_linter.R

```diff
@@ -0,0 +1,841 @@
+test_that("Lint return on end of function", {
+  expect_lint(
+    trim_some("
+      function() {
+        return(1)
+        # Test
+        3 + 4
+      }
+    "),
+    list(
+      line_number = 4L,
+      message = rex::rex("All functions must have an explicit return().")
+    ),
+    return_linter(use_implicit_returns = FALSE)
+  )
+
+  expect_lint(
+    trim_some("
+      function() {
+        return(1)
+      }
+    "),
+    list(
+      line_number = 2L,
+      message = rex::rex("Use implicit return behavior; explicit return() is not needed.")
+    ),
+    return_linter()
+  )
+})
+
+test_that("Lint return on end of lambda function", {
+  skip_if_not_r_version("4.1.0")
+
+  expect_lint(
+    trim_some("
+      \\(bar) {
+        5L + 3L
+      }
+    "),
+    list(
+      line_number = 2L,
+      message = rex::rex("All functions must have an explicit return().")
+    ),
+    return_linter(use_implicit_returns = FALSE)
+  )
+
+  expect_lint(
+    trim_some("
+      \\(bar) {
+        5L + 3L
+        return(1)
+      }
+    "),
+    list(
+      line_number = 3L,
+      message = rex::rex("Use implicit return behavior; explicit return() is not needed.")
+    ),
+    return_linter()
+  )
+})
+
+test_that("Do not lint control statements (with return) on end of function", {
+  linter <- return_linter(use_implicit_returns = FALSE)
+
+  expect_lint(
+    trim_some("
+      function() {
+        repeat {
```

Comment:
@MEO265 shouldn't this lint? Otherwise we're treating `while` and `repeat` differently right?

Currently I'm thinking the rule for explicit returns should be that `for`/`while`/`repeat` shouldn't terminate the function -- better to have an explicit `return(invisible())` come last.